### PR TITLE
implement balance status, fix pre authorize params

### PR DESCRIPTION
**Pull request by illiaredko**
feature/add_balance_status to master

### CODE REVIEW

File: src/main/java/com/icthh/xm/ms/balance/domain/BalanceSpec.java [hunk 1]
```diff
@@ -14,17 +15,21 @@ public class BalanceSpec {
 
     @Data
     public static class BalanceTypeSpec {
+        private static final List<StatusSpec> EMPTY_STATUSES = emptyList();
```
Comment (RomanTeren) [inline src/main/java/com/icthh/xm/ms/balance/domain/BalanceSpec.java]: `emptyList()` itself is a constant, no need to create another variable, just return it in `requireNonNullElse` invocation

File: src/main/java/com/icthh/xm/ms/balance/service/BalanceService.java [hunk 2]
```diff
@@ -611,4 +625,59 @@ private PocketChangeEvent createPocketChangeEvent(Pocket pocket, BigDecimal amou
...
+        if(!balanceStatusKeys.contains(newStatus)){
+            throw new StatusTransitionException(newStatus, null, balanceTypeKey);
+        }
+    }
+
+    private BalanceChangeEvent createChangeStatusChangeEvent(Balance balance, String status) {
+        String operationUuid = UUID.randomUUID().toString();
+        Instant operationDate = now(clock);
+        Map<String, String> metadata = new HashMap<>();
+        metadata.put("change_status_from", balance.getStatus());
```
Comment (RomanTeren) [inline src/main/java/com/icthh/xm/ms/balance/service/BalanceService.java]: let's have some `MetadataConstants` and define these `change_status..` there

File: src/main/java/com/icthh/xm/ms/balance/service/BalanceSpecService.java [hunk 3]
```diff
@@ -93,4 +98,18 @@ public void onInit(String configKey, String configValue) {
     public BalanceSpec.BalanceTypeSpec getBalanceSpec(String typeKey) {
         return getTypeSpecs().getOrDefault(typeKey, NULL_OBJECT);
     }
+    public List<String> balanceStatusKeys(String balanceTypeKey) {
```
Comment (RomanTeren) [inline src/main/java/com/icthh/xm/ms/balance/service/BalanceSpecService.java]: there should be an empty line between methods

Comment (RomanTeren) [inline src/main/java/com/icthh/xm/ms/balance/service/BalanceSpecService.java]: `getBalanceStatusKeys` ?

File: src/main/java/com/icthh/xm/ms/balance/service/BalanceSpecService.java [hunk 4]
```diff
@@ -93,4 +98,18 @@ public void onInit(String configKey, String configValue) {
...
+        List<StatusSpec> statuses = getBalanceSpec(balanceTypeKey).getStatuses();
+        return statuses.stream()
+            .map(StatusSpec::getKey)
+            .collect(Collectors.toList());
+    }
+
+    public Optional<List<NextSpec>> nextSpecs(String balanceTypeKey, String statusKey) {
```
Comment (RomanTeren) [inline src/main/java/com/icthh/xm/ms/balance/service/BalanceSpecService.java]: `getNextStatusSpecs` ?

File: src/main/java/com/icthh/xm/ms/balance/web/rest/BalanceResource.java [hunk 5]
```diff
@@ -103,6 +104,30 @@ public ResponseEntity<BalanceDTO> updateBalance(@Valid @RequestBody BalanceDTO b
             .body(result);
     }
 
+    /**
+     * PUT  /balances : Updates an existing balance status.
```
Comment (RomanTeren) [inline src/main/java/com/icthh/xm/ms/balance/web/rest/BalanceResource.java]: /balances/:id/statuses/:status

File: src/main/java/com/icthh/xm/ms/balance/web/rest/BalanceResource.java [hunk 6]
```diff
@@ -103,6 +104,30 @@ public ResponseEntity<BalanceDTO> updateBalance(@Valid @RequestBody BalanceDTO b
...
+     * or with status 404 (Not Found) if balance not found by id
+     * or with status 500 (Internal Server Error) if the balance couldn't be updated
+     */
+    @PreAuthorize("hasPermission({'id': #id, 'status': #status, 'context': #context}, 'balance', 'BALANCE.STATUS')")
+    @PutMapping("/balances/{id}/statuses/{status}")
+    @Timed
+    @PrivilegeDescription("Privilege to updates an existing balance")
+    public ResponseEntity<BalanceDTO> updateBalanceStatus(@PathVariable Long id,
+                                                          @PathVariable String status,
+                                                          @RequestBody(required = false) Map<String, Object> context){
```
Comment (RomanTeren) [inline src/main/java/com/icthh/xm/ms/balance/web/rest/BalanceResource.java]: pleas reformat all the changes, I've noticed few places with missing spaces

File: src/main/java/com/icthh/xm/ms/balance/domain/BalanceSpec.java (hunk 1 above)
Comment (illiaredko) [reply src/main/java/com/icthh/xm/ms/balance/domain/BalanceSpec.java]: fixed

File: src/main/java/com/icthh/xm/ms/balance/service/BalanceService.java (hunk 2 above)
Comment (illiaredko) [reply src/main/java/com/icthh/xm/ms/balance/service/BalanceService.java]: fixed

File: src/main/java/com/icthh/xm/ms/balance/service/BalanceSpecService.java (hunk 3 above)
Comment (illiaredko) [reply src/main/java/com/icthh/xm/ms/balance/service/BalanceSpecService.java]: fixed

Comment (illiaredko) [reply src/main/java/com/icthh/xm/ms/balance/service/BalanceSpecService.java]: fixed

File: src/main/java/com/icthh/xm/ms/balance/service/BalanceSpecService.java (hunk 4 above)
Comment (illiaredko) [reply src/main/java/com/icthh/xm/ms/balance/service/BalanceSpecService.java]: fixed

File: src/main/java/com/icthh/xm/ms/balance/web/rest/BalanceResource.java (hunk 5 above)
Comment (illiaredko) [reply src/main/java/com/icthh/xm/ms/balance/web/rest/BalanceResource.java]: fixed

File: src/main/java/com/icthh/xm/ms/balance/web/rest/BalanceResource.java (hunk 6 above)
Comment (illiaredko) [reply src/main/java/com/icthh/xm/ms/balance/web/rest/BalanceResource.java]: fixed

File: src/main/java/com/icthh/xm/ms/balance/config/LepContext.java [hunk 7]
```diff
@@ -22,13 +24,15 @@ public class LepContext {
 
     public LepServiceFactory lepServices;
     public LepServices services;
+    public Map<String, Object> templates;
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/ms/balance/config/LepContext.java]: pls replace this map to static class with actual fields with specified classes

Comment (illiaredko) [reply src/main/java/com/icthh/xm/ms/balance/config/LepContext.java]: fixed

**Review by sergeysenja1992 — APPROVED** [review]

**Review by amedvedchuk — APPROVED** [review]

**Pull request merged**